<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Annotate FST/DMR slide windows with genomic features — anno_fst_dmr • GAnnoViz</title><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Annotate FST/DMR slide windows with genomic features — anno_fst_dmr"><meta name="description" content="Annotate FST/DMR slide windows with genomic features."><meta property="og:description" content="Annotate FST/DMR slide windows with genomic features."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAnnoViz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span> Home</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-passport"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-code"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Tutorial.html"><span class="fa fa-file-code"></span> Tutorial</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/benben-miao/GAnnoViz/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotate FST/DMR slide windows with genomic features</h1>
      <small class="dont-index">Source: <a href="https://github.com/benben-miao/GAnnoViz/blob/HEAD/R/anno_fst_dmr.R"><code>R/anno_fst_dmr.R</code></a></small>
      <div class="d-none name"><code>anno_fst_dmr.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate <b><em>FST/DMR</em></b> slide windows with <b><em>genomic features</em></b>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">anno_fst_dmr</span><span class="op">(</span></span>
<span>  <span class="va">gff_file</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  <span class="va">genomic_ranges</span>,</span>
<span>  chrom_col <span class="op">=</span> <span class="st">"CHROM"</span>,</span>
<span>  start_col <span class="op">=</span> <span class="st">"BIN_START"</span>,</span>
<span>  end_col <span class="op">=</span> <span class="st">"BIN_END"</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  ignore_strand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"promoter"</span>, <span class="st">"UTR5"</span>, <span class="st">"gene"</span>, <span class="st">"exon"</span>, <span class="st">"intron"</span>, <span class="st">"CDS"</span>, <span class="st">"UTR3"</span>, <span class="st">"intergenic"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gff-file">gff_file<a class="anchor" aria-label="anchor" href="#arg-gff-file"></a></dt>
<dd><p>Genomic structural annotation <b><code class="file">GFF3/GTF</code></b> file path.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>Format of GFF3/GTF file. (<b><em>"auto"</em></b>, "gff3", "gtf").</p></dd>


<dt id="arg-genomic-ranges">genomic_ranges<a class="anchor" aria-label="anchor" href="#arg-genomic-ranges"></a></dt>
<dd><p>Genomic ranges file (e.g., FST, DMR). (<b><em>chromosome (prefix: chr), start, end</em></b>).</p></dd>


<dt id="arg-chrom-col">chrom_col<a class="anchor" aria-label="anchor" href="#arg-chrom-col"></a></dt>
<dd><p>Chromosome column name. (<b><em>FST: "CHROM", DMR: "chr"</em></b>).</p></dd>


<dt id="arg-start-col">start_col<a class="anchor" aria-label="anchor" href="#arg-start-col"></a></dt>
<dd><p>Start column name. (<b><em>FST: "BIN_START", DMR: "start"</em></b>).</p></dd>


<dt id="arg-end-col">end_col<a class="anchor" aria-label="anchor" href="#arg-end-col"></a></dt>
<dd><p>End column name. (<b><em>FST: "BIN_END", DMR: "end"</em></b>).</p></dd>


<dt id="arg-upstream">upstream<a class="anchor" aria-label="anchor" href="#arg-upstream"></a></dt>
<dd><p>Promoter upstream (bp). (<b><em>2000</em></b>).</p></dd>


<dt id="arg-downstream">downstream<a class="anchor" aria-label="anchor" href="#arg-downstream"></a></dt>
<dd><p>Promoter downstream (bp). (<b><em>200</em></b>).</p></dd>


<dt id="arg-ignore-strand">ignore_strand<a class="anchor" aria-label="anchor" href="#arg-ignore-strand"></a></dt>
<dd><p>Whether to ignore strand when computing overlaps. (<b><em>TRUE</em></b>).</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Which feature types to annotate. c(<b><em>"promoter", "UTR5", "gene", "exon", "intron", "CDS", "UTR3", "intergenic"</em></b>).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Data.frame with input intervals and <b><em>annotation results</em></b>.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>benben-miao</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example GFF3 file in GAnnoViz</span></span></span>
<span class="r-in"><span><span class="va">gff_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"example.gff"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"GAnnoViz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate FST</span></span></span>
<span class="r-in"><span><span class="va">fst_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"example.fst"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"GAnnoViz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">anno_fst_dmr</span><span class="op">(</span></span></span>
<span class="r-in"><span>  gff_file <span class="op">=</span> <span class="va">gff_file</span>,</span></span>
<span class="r-in"><span>  format <span class="op">=</span> <span class="st">"auto"</span>,</span></span>
<span class="r-in"><span>  genomic_ranges <span class="op">=</span> <span class="va">fst_table</span>,</span></span>
<span class="r-in"><span>  chrom_col <span class="op">=</span> <span class="st">"CHROM"</span>,</span></span>
<span class="r-in"><span>  start_col <span class="op">=</span> <span class="st">"BIN_START"</span>,</span></span>
<span class="r-in"><span>  end_col <span class="op">=</span> <span class="st">"BIN_END"</span>,</span></span>
<span class="r-in"><span>  upstream <span class="op">=</span> <span class="fl">2000</span>,</span></span>
<span class="r-in"><span>  downstream <span class="op">=</span> <span class="fl">200</span>,</span></span>
<span class="r-in"><span>  ignore_strand <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"promoter"</span>, <span class="st">"UTR5"</span>, <span class="st">"gene"</span>, <span class="st">"exon"</span>, <span class="st">"intron"</span>, <span class="st">"CDS"</span>, <span class="st">"UTR3"</span>, <span class="st">"intergenic"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM BIN_START BIN_END N_VARIANTS WEIGHTED_FST    MEAN_FST</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 chr11     10001   20000          7   0.04366440  0.02391140</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 chr11     15001   25000          7   0.04366440  0.02391140</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 chr11     25001   35000          4   0.01503910  0.01619730</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 chr11     30001   40000          4   0.01503910  0.01619730</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 chr11     55001   65000          1  -0.00444874 -0.00444874</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 chr11     60001   70000          1  -0.00444874 -0.00444874</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 anno_type                   gene_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 promoter,gene,UTR5,CDS,exon,intron,UTR3 HdF041849(p),HdF041849(g)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          promoter,gene,UTR5,exon,intron HdF041849(p),HdF041849(g)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                              intergenic                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                              intergenic                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                              intergenic                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                              intergenic                          </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate DMR</span></span></span>
<span class="r-in"><span><span class="va">dmr_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"example.dmr"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"GAnnoViz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">anno_fst_dmr</span><span class="op">(</span></span></span>
<span class="r-in"><span>  gff_file <span class="op">=</span> <span class="va">gff_file</span>,</span></span>
<span class="r-in"><span>  format <span class="op">=</span> <span class="st">"auto"</span>,</span></span>
<span class="r-in"><span>  genomic_ranges <span class="op">=</span> <span class="va">dmr_table</span>,</span></span>
<span class="r-in"><span>  chrom_col <span class="op">=</span> <span class="st">"chr"</span>,</span></span>
<span class="r-in"><span>  start_col <span class="op">=</span> <span class="st">"start"</span>,</span></span>
<span class="r-in"><span>  end_col <span class="op">=</span> <span class="st">"end"</span>,</span></span>
<span class="r-in"><span>  upstream <span class="op">=</span> <span class="fl">2000</span>,</span></span>
<span class="r-in"><span>  downstream <span class="op">=</span> <span class="fl">200</span>,</span></span>
<span class="r-in"><span>  ignore_strand <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"promoter"</span>, <span class="st">"UTR5"</span>, <span class="st">"gene"</span>, <span class="st">"exon"</span>, <span class="st">"intron"</span>, <span class="st">"CDS"</span>, <span class="st">"UTR3"</span>, <span class="st">"intergenic"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Each of the 2 combined objects has sequence levels not in the other:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'x': h1tg000177l, h1tg000356l, h1tg000377l, h1tg000386l, h1tg000411l, h1tg000632l, h1tg000725l, h1tg000791l, h1tg000855l, h1tg000859l, superscaf9, h1tg000571c, h1tg000573l, h1tg000597l, h1tg000610l, h1tg000665l, h1tg000675l, h1tg000822l</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   - in 'y': h1tg000310c, h1tg000374l, h1tg000441l, h1tg000520l, h1tg000541l, h1tg000549l, h1tg000554l, h1tg000609l, h1tg000618l, h1tg000620l, h1tg000625l, h1tg000633l, h1tg000686l, h1tg000739l, h1tg000742l, h1tg000750l, h1tg000756l, h1tg000761l, h1tg000769l, h1tg000774l, h1tg000782l, h1tg000790l, h1tg000797l, h1tg000798l, h1tg000799l, h1tg000831l, h1tg000840l, h1tg000847l, h1tg000852l, h1tg000856l, h1tg000864l, h1tg000873l, h1tg000875l, h1tg000878l, h1tg000890l, h1tg000906l, h1tg000912l, h1tg000913l, mith1tg000589c, superscaf10, superscaf13, superscaf27, superscaf28, superscaf29, superscaf3, superscaf31, superscaf32, superscaf33, superscaf35, superscaf37, superscaf38, superscaf40</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Make sure to always combine/compare objects based on the same reference</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   genome (use suppressWarnings() to suppress this warning).</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    chr   start     end strand        pvalue        qvalue meth.diff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 chr1  466001  468000      *  4.157118e-03  1.206712e-02 -27.31707</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 chr1  660001  662000      *  4.041133e-09  3.622605e-08 -33.37925</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 chr1 1454001 1456000      *  8.939178e-12  1.083862e-10 -26.55631</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 chr1 2750001 2752000      * 2.716109e-129 5.710428e-127  25.04745</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 chr1 3428001 3430000      *  2.049072e-10  2.144328e-09 -46.33124</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 chr1 3604001 3606000      *  2.137854e-09  1.984104e-08  37.85743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        anno_type      gene_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    gene,intron HdF051575(g)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    gene,intron HdF051575(g)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 gene,exon,UTR3 HdF027216(g)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     intergenic             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    gene,intron HdF027234(g)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    gene,intron HdF027242(g)</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Benben Miao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

