<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Annotate differentially expressed genes (DEGs) with chromosome positions — anno_deg_chrom • GAnnoViz</title><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/JetBrains_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Annotate differentially expressed genes (DEGs) with chromosome positions — anno_deg_chrom"><meta name="description" content="Annotate differentially expressed genes (DEGs) with chromosome positions."><meta property="og:description" content="Annotate differentially expressed genes (DEGs) with chromosome positions."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAnnoViz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span> Home</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-passport"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-file-code"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Tutorial.html"><span class="fa fa-file-code"></span> Tutorial</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/benben-miao/GAnnoViz/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Annotate differentially expressed genes (DEGs) with chromosome positions</h1>
      <small class="dont-index">Source: <a href="https://github.com/benben-miao/GAnnoViz/blob/HEAD/R/anno_deg_chrom.R"><code>R/anno_deg_chrom.R</code></a></small>
      <div class="d-none name"><code>anno_deg_chrom.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate <b><em>differentially expressed genes (DEGs)</em></b> with chromosome positions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">anno_deg_chrom</span><span class="op">(</span></span>
<span>  <span class="va">deg_file</span>,</span>
<span>  <span class="va">gff_file</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  id_col <span class="op">=</span> <span class="st">"GeneID"</span>,</span>
<span>  fc_col <span class="op">=</span> <span class="st">"log2FoldChange"</span>,</span>
<span>  use_strand <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  drop_unmapped <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-deg-file">deg_file<a class="anchor" aria-label="anchor" href="#arg-deg-file"></a></dt>
<dd><p>DEG table from <b><em>DESeq2</em></b> analysis.</p></dd>


<dt id="arg-gff-file">gff_file<a class="anchor" aria-label="anchor" href="#arg-gff-file"></a></dt>
<dd><p>Genomic structural annotation <b><code class="file">GFF3/GTF</code></b> file path.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>Format of GFF3/GTF file. (<b><em>"auto"</em></b>, "gff3", "gtf").</p></dd>


<dt id="arg-id-col">id_col<a class="anchor" aria-label="anchor" href="#arg-id-col"></a></dt>
<dd><p>Gene IDs column name. (<b><em>"GeneID"</em></b>).</p></dd>


<dt id="arg-fc-col">fc_col<a class="anchor" aria-label="anchor" href="#arg-fc-col"></a></dt>
<dd><p>Log2(fold change) column name. (<b><em>"log2FoldChange"</em></b>).</p></dd>


<dt id="arg-use-strand">use_strand<a class="anchor" aria-label="anchor" href="#arg-use-strand"></a></dt>
<dd><p>Whether to resput actual strand instead of <code>"*"</code>. (<b><em>FALSE</em></b>).</p></dd>


<dt id="arg-drop-unmapped">drop_unmapped<a class="anchor" aria-label="anchor" href="#arg-drop-unmapped"></a></dt>
<dd><p>Whether to drop DEG entries not found in annotation. (<b><em>TRUE</em></b>).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame with columns: (<b><em>chrom, start, end, GeneID, log2FoldChange, strand</em></b>).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example DEGs and GFF in GAnnoViz</span></span></span>
<span class="r-in"><span><span class="va">deg_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"example.deg"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"GAnnoViz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gff_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"example.gff"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"GAnnoViz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Annotate DEGs with chromosome positions</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">anno_deg_chrom</span><span class="op">(</span></span></span>
<span class="r-in"><span>  deg_file <span class="op">=</span> <span class="va">deg_file</span>,</span></span>
<span class="r-in"><span>  gff_file <span class="op">=</span> <span class="va">gff_file</span>,</span></span>
<span class="r-in"><span>  format <span class="op">=</span> <span class="st">"auto"</span>,</span></span>
<span class="r-in"><span>  id_col <span class="op">=</span> <span class="st">"GeneID"</span>,</span></span>
<span class="r-in"><span>  fc_col <span class="op">=</span> <span class="st">"log2FoldChange"</span>,</span></span>
<span class="r-in"><span>  use_strand <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  drop_unmapped <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chrom    start      end   gene_id     score strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 chr14 12802589 12807112 HdF000031  1.408356      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 chr14 12979855 12984114 HdF000041 -1.114777      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 chr14 13092225 13094734 HdF000049  1.271638      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 chr14 13915421 13919063 HdF000066 -6.180426      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 chr14 15038730 15050966 HdF000091 -1.250090      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 chr14 15455652 15456778 HdF000106 -6.085891      *</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Benben Miao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

